### ISSUE TITLE
Cannot start another process while you are in the child process at 

### ISSUE THREAD

**Issue by thegamecat** [xFFFFF/Gekko-BacktestTool#5]
perl backtest.pl
Cannot start another process while you are in the child process at /usr/local/share/perl/5.22.1/Parallel/ForkManager.pm line 44.
Cannot start another process while you are in the child process at /usr/local/share/perl/5.22.1/Parallel/ForkManager.pm line 44.
Cannot start another process while you are in the child process at /usr/local/share/perl/5.22.1/Parallel/ForkManager.pm line 44.
Cannot start another process while you are in the child process at /usr/local/share/perl/5.22.1/Parallel/ForkManager.pm line 44.
Cannot start another process while you are in the child process at /usr/local/share/perl/5.22.1/Parallel/ForkManager.pm line 44.

**Comment by xFFFFF** [comment]
Its propablly fixed in v0.4. Im close.

**Comment by bald123** [comment]
I get the exact same results with the following parameters (v0.7, running on raspbian): 

@strategies = qw(
MACD
CCI
RSI
);

@pairs = qw(
kraken:EUR:XRP
);

@warmup = qw(
15:1
30:1
60:1
);

$from = '2018-02-01 00:00:00';
$to = '2018-02-05 00:00:00';

I did not manage to get this error before (yesterday).

**Comment by xFFFFF** [comment]
Its bug. Use only one value in @warmup

**Comment by bald123** [comment]
I only seem to get this error when I don't install the Gekko-Strategies. When I do install them, I don't see this specific error (when using multiple warm-ups), but I don't get any results as well. I'm going to try now with one warm-up value.

**Comment by xFFFFF** [comment]
Strategies havent nothing to do with BT :) If it's not a mistake, it seems strange.

What do You mean in last question?

**Comment by brunoaduarte** [comment]
Nice, it works with only one value in @WarmUp.
Problem is that the results from gekko interface are not the same as Gekko-BacktestTool

I did one test with the following parameters:

- From 01/01/2018 00:00 to 22/05/2018 23:59
- Only 1 indicator -> RSI (with default parameters)
- Binance - BTC/USDT pair
- 15 min candle / 10 period warmup
- Customized paper trader settings (same on both tests)

## Here's the results with Gekko interface:

![image](https://user-images.githubusercontent.com/5354345/40517485-efad4930-5f8b-11e8-9dab-506e1fdd4cf8.png)

**Market -41,80% / Simulated profit -56,36%**

## And results with Gekko-BacktestTool
```

                                                                                      █████               ████
                                                                                      █████               ████
           ██                 ██                                                      █████               ████
           ██  █████████████  ██                                                      █████               ████
           █████████████████████                 █████████████        █████████       █████      █████    ████      █████      █████████
           █████████████████████               ███████████████     ██████████████     █████    █████      ████    ██████     ██████████████
           █████████████████████              █████      █████    █████      █████    █████  █████        ████  ██████      █████      █████
         █████████████████████████           █████       █████   █████        ████    █████ █████         ██████████       █████        █████
         █████████████████████████           ████        █████   ██████████████████   █████████           ████████         ████         █████
         ██████    ██████    █████           ████        █████   ██████████████████   ███████████         ██████████       ████         █████
           ████   ███████   ████             █████       █████   █████                █████  █████        ████  ██████     █████        █████
            ███████████████████               ████████████████    ██████              █████   ██████      ████    █████     █████      █████
             █████████████████                  ██████████████     ██████████████     █████     ██████    ████     █████     ██████████████
               █████████████                             █████        ███████████     █████       █████   ████       █████      ████████
                   ██████                                ████
                                              ███████████████
                                              █████████████


                 ████████                      ███         ███                            ██     █████████                     █████
                 ███   ███   ████              ███         ███                  ██        ██        ███                           ██
                 ████████   █    ██   ███████  ███ ███   ████████   ███████   ██████   █████████    ███     ███████   ███████     ██
                 ████████    ███████  ██       ██████      ███     █████████  █████       ██        ███     ██   ███  ██   ███    ██
                 ███   ███ ███   ███  ██       ███ ███     ███     ███            ███     ██        ███     ██   ███  ██   ███    ██
                 █████████ █████████   ██████  ███  ███     █████   ████████ ████████     ██████    ███     ███████   ███████     ██████

                                                                                                        Gekko BacktestTool v0.7
                                                                                              https://github.com/xFFFFF/Gekko-BacktestTool


[2018-05-24 22:56:28] Collect data from coinmarketcap.com
[2018-05-24 22:56:28] Use TOML files for strategies configuration
[2018-05-24 22:56:28] Save output to csvfile: database.csv
[2018-05-24 22:56:28] Starting Backtest Machine (5 threads)...
============================================================== I N P U T    D A T A ===============================================================
Pairs
-----
binance:USDT:BTC
Total: 1
===================================================================================================================================================
Strategies
----------
RSI
Total: 1

===================================================================================================================================================
Candles
-------
15:10
Total: 1
===================================================================================================================================================
[2018-05-24 22:56:29] 1 backtests remain
============================================================= L A S T    R E S U L T S ============================================================
Pair                                      Strategy       Profit   Profit-market  Trades/day  Win trades  Best win  Worst loss    HODL  Days   Time
----                                      --------       ------   -------------  ----------  ----------  --------  ----------    ----  ----   ----
USDT:BTC                                       RSI      -70.14%          -28.34        0.60      52.38%     7.08%     -20.33%    2608   141  00:10
===================================================================================================================================================

[2018-05-24 22:56:39] 100% complete. Elapsed: 00:00:10, average backtest duration: 00:02


[2018-05-24 22:56:39] Creating ALL RESULTS table (sorted profit/day)...

==================================================== A L L     R E S U L T S =====================================================
Curr. Asset Strat Profit Profit-market   Profit/day Trades/day Best win Worst loss Avg.Hodl     Avg.price Ov.trades/day Volume/day
----  ----- ----- ------ ------------- ------------ ---------- -------- ---------- --------     --------- -------------  ---------
USDT  BTC   RSI   -70.14        -28.34        -0.50       0.60     7.08     -20.33     2608 9675.09656254        248196      35536
==================================================================================================================================


[2018-05-24 22:56:39] All jobs are done. Elapsed time: 11 seconds
```
**backtest-config.pl**
```
no warnings qw(uninitialized);
############################# START OF CONFIGURATION #############################
# Put your strategy names between brackets in line below. Strategy seperate with space or newline. You can add all Your strategies from gekko/strategies directory with adding line ALL.
@strategies = qw(
RSI
);
# Put your pairs between brackets in line below. Use exchange:currency:asset format. Seperate pair using space or newline. You can add all Your paris with ALL line or all pairs for exchange with binance:ALL line or binance:USDT:ALL line. Another option is adding dates for dataset for indivual pairs, ex: binance:BNB:NULS:2018-04-05:2018-05-01
@pairs = qw(
binance:USDT:BTC
);

# Put your candle values between brackets in line below. Use CandleSize:WarmupPeriod format. Seperate pair using space or newline.
@warmup = qw(
15:10
);

############################# OPTIONAL SETTINGS #############################
# To specify time range for import or backtest uncomment lines below, but instead this you can use command line input ex.: backtest.pl --from "2018-01-01 00:00:00" --to "2018-01-05 00:00:00". If below lines are commented Gekko is using scan datasets feature in backtest mode.
$from = '2018-01-01 00:00:00';
$to = '2018-05-22 23:59:00';

# If You are using only one exchange or one exchange and one currency You can put default values below, and adding only asset name to @pairs ex: NULS, ADA, TRX - without binance:BTC before asset ex: perl backtest.pl -p NULS,ADA,TRX.
#$default_set = 'binance:BNB';

# CSV file name. You don't need change this. All new data will append to exist file without deleting or replacing.
$csv = 'database.csv';

# You can add note to project below. Note will be add in CSV file. Its can be useful when You are developing strategy.
$note = '';

# Sort strategies in top list by. Available values to sort: best, profitable, profit_above_market, best_PL, worst_PL, profits_sum, avg_profit, trades_win, trades_day, hodl_time
$top_strategy_sort1 = 'best';
$top_strategy_sort2 = 'profitable';

# Sort datasets ranking by. Available values to sort: best, profitable, profit_above_market, market_change, best_PL, worst_PL, profits_sum, avg_profit, trades_win, trades_day, hodl_time, price_volatility, cmc_rank, cmc_marketcap, cmc_volume, days
$top_dataset_sort1 = 'best';
$top_dataset_sort2 = 'profitable';

# Template of CSV output columns. Format [% variable_name %], columns MUST be seperated by comma (,) without any space.
# Below is compact version
$csv_columns = \ "[% currency %],[% asset %],[% strategy %],[% profit %],[% profit_market %],[% profit_day %],[% market_change %],[% trades_day %],[% percentage_wins %],[% best_win %],[% median_wins %],[% worst_loss %],[% median_losses %],[% avg_exposed_duration %],[% candle_size %],[% warmup_period %],[% dataset_days %],[% CMC_Rank %],[% current_marketcap %],[% open_price %],[% close_price %],[% lowest_price %],[% highest_price %],[% avg_price %],[% price_volality %],[% volume_day %],[% volume_CMC %],[% overall_trades_day %],[% dataset_from %],[% dataset_to %],[% strategy_settings %],[% note %]";
# Minimalistic version - tables will dont generate
# $csv_columns = \ "[% currency %],[% asset %],[% strategy %],[% profit %],[% trades_day %],[% percentage_wins %],[% best_win %],[% worst_loss %],[% avg_exposed_duration %],[% median_wins %],[% median_losses %],[% dataset_from %],[% dataset_to %],[% strategy_settings %],[% profit_day %],[% profit_market %],[% avg_price %],[% price_volality %],[% volume_day %],[% volume_CMC %],[% CMC_Rank %],[% current_marketcap %],[% overall_trades_day %],[% dataset_days %],[% market_change %]";
# Full version - all possible BacktestTool variables.
#$csv_columns = \ "[% currency %],[% asset %],[% exchange %],[% strategy %],[% profit %],[% profit_day %],[% profit_year %],[% sharpe_ratio %],[% market_change %],[% profit_market %],[% trades %],[% trades_day %],[% winning_trades %],[% lost_trades %],[% percentage_wins %],[% best_win %],[% median_wins %],[% worst_loss %],[% median_losses %],[% avg_exposed_duration %],[% candle_size %],[% warmup_period %],[% dataset_days %],[% backtest_start %],[% dataset_from %],[% dataset_to %],[% CMC_Rank %],[% current_marketcap %],[% open_price %],[% close_price %],[% lowest_price %],[% highest_price %],[% avg_price %],[% price_volality %],[% volume %],[% volume_day %],[% volume_CMC %],[% overall_trades %],[% overall_trades_day %],[% note %]";

# Do You want coinmarketcap.com data in CSV output?
$cmc_data = 'yes';

# Single backtest results interval for print backtests summary (% complete, eta, avg backtest time, elapsed) for all results from current instance
$summary_interval = 60;
# Print above tables on each summary?
$print_analysis_on_summary = 'yes';

# Print ALL RESULTS table when use command backtest -a CSV file and at end of all backtests?
$print_all_results = 'yes';
# Minimum and maximum values to appear in the table with all results. Results that do not meet the following values will not be printed.
$all_results_min_profit = -999999999;
$all_results_min_profit_market = -99999999999;
$all_results_min_profit_day = -9999;
$all_results_min_trades_day = 0;
$all_results_max_trades_day = 99999;
$all_results_min_hodl_time = 1;
$all_results_max_hodl_time = 99999;
# How many rows from ALL RESULTS would You print in terminal output?
$all_results_limit = 100;

# Print TOP STRATEGY table when use command backtest -a CSV file and at end of all backtests?
$print_top_strategy = 'no';
# Sort strategies in top list by. Available values to sort: best, profitable, profit_above_market, best_PL, worst_PL, profits_sum, avg_profit, trades_win, trades_day, hodl_time
$top_strategy_sort1 = 'trades_win';
$top_strategy_sort2 = 'avg_profit';
# Limit for printed results in TOP STRATEGY table
$top_strategy_limit = 100;

# Print TOP DATASET table when use command backtest -a CSV file and at end of all backtests?
$print_top_dataset = 'no';
# Sort datasets ranking by. Available values to sort: best, profitable, profit_above_market, market_change, best_PL, worst_PL, profits_sum, avg_profit, trades_win, trades_day, hodl_time, price_volatility, cmc_rank, cmc_marketcap, cmc_volume, days
$top_dataset_sort1 = 'trades_win';
$top_dataset_sort2 = 'best_PL';
# Limit for printed results in TOP DATASET table
$top_dataset_limit = 100;

# Do you want see roundtrips report in terminal output?
$print_roundtrips = 'no';

# Use TOML strat's config files instead JSON?
$use_toml_files = 'yes';
$toml_directory = 'config/strategies/';

# Do you need Gekko's log files in log directory?
$keep_logs = 'no';

# Threads amount, for 4xcpu cores is recommended to set 5-6 value.
$threads = 5;

# When you set stfu to 'yes' only results will be displayed.
$stfu = 'no';

# If You set $use_toml_files to 'no' then add Your strat's config in JSON format between brackets below.
#$stratsettings = q(
#config.neuralnet = {
#  "threshold_buy": 1,
#  "threshold_sell": -1,
#  "price_buffer_len": 100,
#  "learning_rate": 1,
#  "momentum": 0.9,
#  "decay": 0.01,
#  "min_predictions": 1000,
#  "stoploss_enabled": false,
#  "stoploss_threshold": 0.95
#};

#config.neuralnet_BULL_BEAR = {
#  "threshold_sell_bear":-1.0352,
#  "threshold_sell_bull":-0.992,
#  "momentum":0.0982,
#  "decay":0.0076,
#  "threshold_buy_bull":1.955,
#  "price_buffer_len":88.6,
#  "SMA_short":38,
#  "SMA_long":712,
#  "min_predictions":760,
#  "threshold_buy_bear":2.8294,
#  "learning_rate":0.2736
#};

#);

# Other Gekko's settings for backtest
$asset_c = 1;
$currency_c = 0;
$fee_maker = 0.1;
$fee_taker = 0.1;
$fee_using = 'maker';
$slippage = 0.5;
$riskFreeReturn = 5;
############################# END OF CONFIGURATION #############################

1;
```

Maybe i'm doing something wrong here ?

**Comment by xFFFFF** [comment]
Check wiki please ;-)

**Comment by bald123** [comment]
Yes of course strategies have nothing to do with it. It was late already and was not thinking clearly anymore. 

What I meant with "to re-open this bug": right now this bug is closed on github (this very page where we're looking at), while it is still present in the software. Since the power of the backtesttool lies in the fact that multiple strategies and config settings can be tested quick and easily, it is kind of a showstopper when only one candle configuration can be used. Once a bug is closed, but still present, it should appear again under the "Open issues" so that people know that it is still present (and that it still requires some work).

**Comment by bald123** [comment]
This issue seems to be fixed when the following line on rule 1183 is changed from:
if (@pairs > scalar (keys %stratconfig) && @pairs >= @warmup && @pairs > 1 || $bftriger) {

to

if (@pairs >= scalar (keys %stratconfig) && @pairs >= @warmup && @pairs > 1 || $bftriger) {

Change the ">" to ">=". I did no extensive testing yet.

**Comment by bald123** [comment]
Is there anything I can do to help fix this issue? I'm afraid I have no python experience though.

**Comment by bald123** [comment]
I seem to have it fixed temporarily (tested with combo of 3 pairs, 3 warmup periods and 3 strategies in backtest-config.pl). I only fork a new process at the most inner iteration, instead of at each iteration (warmups, strategies, pairs). In my theory this should not have a negative effect on the performance, it might even be a little bit quicker because less processes are being forked. The real "processing" happens in the most inner iteration.

Change the piece of code where the iterations start (commented out some lines):

  foreach (@warmup) {
    #if (@warmup >= scalar (keys %stratconfig) && @warmup >= @pairs && @warmup > 1) {
      #$pid = $pm->start and next;
    #}
    my @warms = split /:/, $_;
    foreach (@pairs) {
      #if (@pairs > scalar (keys %stratconfig) && @pairs >= @warmup && @pairs > 1 || $bftriger) {
      #  $pid = $pm->start and next;
      #}
      my @sets = split /:/, $_;
      foreach my $stratn (keys %stratconfig) {
        #print "\n\n+++++++++++++++++++++++++++++++++\nDEBUG\nStrat name: $stratn \nConfig: \n$stratconfig{$stratn}\n++++++++++++++++++++++++++++++++++++++++++\n\n";
        #if (scalar (keys %stratconfig) >= @pairs && scalar (keys %stratconfig) >= @warmup && scalar (keys %stratconfig) > 1 && ! defined $bftriger) {
          $pid = $pm->start and next;
        #}

And at the end of the iterations:

        #if (scalar (keys %stratconfig) >= @pairs && scalar (keys %stratconfig) >= @warmup && scalar (keys %stratconfig) > 1 && ! $bftriger) {
          $pm->finish(0, \%paircalc);
        #}
      }
      #if (@pairs >= scalar (keys %stratconfig) && @pairs >= @warmup && @pairs > 1 || $bftriger) {
      #  $pm->finish(0, \%paircalc);
      #}
    }
    #if (@warmup >= scalar (keys %stratconfig) && @warmup >= @pairs && @warmup > 1) {
    #  $pm->finish(0, \%paircalc);
    #}


Haven't made a pull request yet, because this is just a temporary hack from a developer who has never touched python before (I also haven't investigated "\%paircalc" yet)... Filip, don't you think that this might improve performance since process forking might be more expensive than the workload actually is in the first two iterations?
